rspotify-model/src/idtypes.rs

```diff
@@ -296,9 +224,243 @@ impl<T: IdType> Id<T> {
     }
+
+    /// Parse Spotify id or URI from string slice
+    ///
+    /// Spotify URI must be in one of the following formats:
+    /// `spotify:{type}:{id}` or `spotify/{type}/{id}`.
+    /// Where `{type}` is one of `artist`, `album`, `track`, `playlist`, `user`,
+    /// `show`, or `episode`, and `{id}` is a non-empty alphanumeric string.
+    /// The URI must be match with the ID's type (`Id::TYPE`), otherwise
+    /// `IdError::InvalidType` error is returned.
+    ///
+    /// Examples: `spotify:album:6IcGNaXFRf5Y1jc7QsE9O2`,
+    /// `spotify/track/4y4VO05kYgUTo2bzbox1an`.
+    ///
+    /// If input string is not a valid Spotify URI (it's not started with
+    /// `spotify:` or `spotify/`), it must be a valid Spotify object id,
+    /// i.e. a non-empty alphanumeric string.
+    ///
+    /// # Errors:
+    ///
+    /// - `IdError::InvalidType` - if `id_or_uri` is an URI, and it's type part
+    ///    is not equal to `T`,
+    /// - `IdError::InvalidId` - either if `id_or_uri` is an URI with invalid id
+    ///    part, or it's an invalid id (id is invalid if it contains
+    ///    non-alphanumeric characters),
+    /// - `IdError::InvalidFormat` - if `id_or_uri` is an URI, and it can't be
+    ///    split into type and id parts.
+    fn from_id_or_uri(id_or_uri: &str) -> Result<Self, IdError>
+    where
+        Self: Sized,
+    {
+        match Self::from_uri(id_or_uri) {
+            Ok(id) => Ok(id),
+            Err(IdError::InvalidPrefix) => Self::from_id(id_or_uri),
+            Err(error) => Err(error),
+        }
+    }
+}
+
+pub trait PlayableId: Id {}
+pub trait PlayContextId: Id {}
+
+/// This macro helps consistently define ID types. It contains a lot of code but
+/// mostly it's just repetitive work that's not of much interest for being
+/// trivial.
+///
+/// * The `$type` parameter indicates what type the ID is made out of (say,
+///   `Artist`, `Album`...) from the enum `Type`.
+/// * The `$name` parameter is the identifier of the struct for that type.
+macro_rules! define_idtypes {
+    ($($type:ident => $name:ident),+) => {
+        $(
+            #[doc = concat!(
+                "ID of type [`Type::",
+                stringify!($type),
+                "`]. Refer to the [module-level docs][`crate::idtypes`] for more information.")
+            ]
+            #[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize, Hash)]
+            pub struct $name(String);
+
+            impl Id for $name {
+                #[inline]
+                fn id(&self) -> &str {
+                    &self.0
+                }
+
+                #[inline]
+                fn _type(&self) -> Type {
+                    Type::$type
+                }
+
+                #[inline]
+                fn _type_static() -> Type where Self: Sized {
+                    Type::$type
+                }
+
+                #[inline]
+                unsafe fn from_id_unchecked(id: &str) -> Self {
+                    $name(id.to_owned())
```

Comment:
I am a little unsure that is it possible to construct a `$name`(for example `AlbumId`) from `String`? It seems there is no such constructor?

Reply:
Yes, it's a [tuple struct](https://learning-rust.github.io/docs/b2.structs.html#Tuple-structs) (defined as `struct $name(String)`), which you can initialize with `$name(my_string)`